### ISSUE TITLE
PB Banana struggles transmitting large ints

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2125]
|<img alt="engie's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| engie reported|
|-|-|
|Trac ID|trac#2125|
|Type|defect|
|Created|2006-09-25 15:53:51Z|

Tried to send a large int through PB. Discussed with dash on IRC:

```
(03:52:36 PM) dash: engie: my first guess is that ginormous integer.
(03:52:55 PM) engie: It's a hash (as produced by pycrypto)
(03:53:36 PM) ***dash nods
(03:53:55 PM) jml_ left the room (quit: Read error: 145 (Connection timed out)).
(03:56:09 PM) dash: that is indeed the problem.
(03:57:24 PM) engie: any way around it?
(03:57:39 PM) engie: or is it a pb bug that should be fixeD?
(03:57:48 PM) dash: definitely a bug that should be fixed
(03:57:54 PM) dash: there may be a way around it... hm
(03:57:57 PM) dash: well, heck
(03:58:31 PM) dash: at the very worst, you can just do str(mastersig)
(03:59:37 PM) exarkun: the base 10 representation of an integer which exceeds bananas 64 byte limitation is going to be... long
(04:00:00 PM) STemplar left the room (quit: Read error: 110 (Connection timed out)).
(04:01:02 PM) dash: >>> len(str(massiveInteger))
(04:01:02 PM) dash: 923
(04:02:07 PM) dialtone [n=dialtone@...] entered the room.
(04:03:11 PM) dash: not even 1k!
(04:03:40 PM) engie: hehe. How is it tripping up the banana?
(04:03:45 PM) exarkun: oh
(04:03:49 PM) exarkun: that really _is_ a bug
(04:04:05 PM) exarkun: that limit totally shouldn't apply to integers...
(04:04:12 PM) dash: indeed.
```

Attachments:

* [demo.tar.gz](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e8e/e8e44b0fdc368e9421252c098358e3cbd5d353c1/2588b8968a3cc4076904d4c783a616a642e38c99.gz) (61635 bytes) - added by engie on 2006-09-25 15:54:40Z - A test cast. Run twistd -ny app.tac to replicate the bug.

<details><summary>Searchable metadata</summary>

```
trac-id__2125 2125
type__defect defect
reporter__engie engie
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__ 
time__1159199631000000 1159199631000000
changetime__1166850172000000 1166850172000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(Reformatting description)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

I guess this is addressed in banana-longint-2125.  The limit is left at 2 ** 448, but made configurable via the module-level setPrefixLimit function and the Pynana.setPrefixLimit method.

Open questions:
 * Should Canana have a setPrefixLimit?
 * Should the encode and decode methods respect the current module-level prefix limit, or the limit at module-import time (since they re-use a global Banana instance, they'll get whatever settings were present at import time, and never change)
 * Should the default limit be raised from 2 ** 448 - 1 to something more reasonable?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @warner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

I merged this forward to banana-longint-2125-2 because there were some odd problems which might be fixed by doing this.

As long as I'm here, here are my answers to the above open questions:

  * No.  Canana is unused, it's not worth the effort to update.  If anything we should just remove it.
  * The encode and decode methods aren't really used for anything serious.  If you want to do anything serious, you can create a new Banana instance and call methods on it.
  * 2 ** 448 - 1 is pretty big.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_6"></a>|@jerub set owner to @exarkun|
|-|-|

b1282int needs a docstring, and a better name. Please name it something that doesn't parse as "B 1282 Int".

setPrefixLimit's docstring doesn't use a @param, perhaps one is appropriate.

Pynana.setPrefixLimit should have a similar docstring to setPrefixLimit.

Thankyou for the tests and changing all the exceptions around to return exception instances.

Pyflakes warnings:
```
twisted/spread/banana.py:24: 'types' imported but unused
twisted/spread/banana.py:127: undefined name '_PREFIX_LIMIT'
twisted/spread/banana.py:322: undefined name 'cBanana'
twisted/spread/banana.py:323: undefined name 'cBanana'
twisted/spread/banana.py:328: undefined name 'cBanana'
twisted/spread/banana.py:334: undefined name 'cBanana'
```
Recommend fixing first warning, but I'm unconcerned by the rest.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @jerub|
|-|-|

Thanks.

The b1282int optimizations weren't intended to be included in this branch.  I've removed them and created a separate ticket for that.

Fixed up the other issues you mentioned.  I'm not sure what you meant be "changing all the exceptions around to return exception instances", though.

Taking another look at this branch now, I'm trying to remember why I named _getLargest as I did.  I think I will try to remember some more and add a docstring to both it and _getSmallest once I do.  Other than that, I think it's probably ready to be merged.  Take another look?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @jerub|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_10"></a>|@jerub set owner to @exarkun|
|-|-|

Yay! it's all beautiful. Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [18960]) Merge banana-longint-2125-2

Author: exarkun
Reviewer: jerub
Fixes [#2125](https://github.com/twisted/twisted/issues/2125)

Parameterize the prefix limit for long integers and add checking code to
the encode function instead of letting it generate output which will
most likely be rejected during decoding.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

(In [18963]) Revert r18960 due to buildbot failures.  Refs [#2125](https://github.com/twisted/twisted/issues/2125)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph set status to `new`|
|-|-|

If this code is going to make ```cBanana``` a test-failing dud, I'd say it should remove all references to ```cBanana``` - not just the build - as part of the change.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun removed owner|
|-|-|

I removed cBanana and all references to it I could find, except for those in the historic documentation.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph set owner to @exarkun<br>@glyph set status to `new`|
|-|-|

Looking good.  I think this is ready for another merge.

One thing though before you do - there is some trailing whitespace right smack in the middle of the new tests, on line 122 of ```test_banana.py```.  Also, right after the changed ```encClass``` attribute.

Can you just run whitespace-cleanup on that file as part of this branch, since both of those show up in the diff (even though they're not new)?  If you have any reason not to, go ahead and merge without.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set status to `closed`|
|-|-|

(In [19014]) Merge banana-longint-2125-2

Author: exarkun
Reviewer: jerub, glyph
Fixes [#2125](https://github.com/twisted/twisted/issues/2125)

Parameterize the prefix limit for long integers and add checking code to
the encode function instead of letting it generate output which will
most likely be rejected during decoding.

Also remove cBanana, since it is unused and no longer passes all of
the test Banana tests.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
